### PR TITLE
feat: unsafe and quoteIdent

### LINKED ISSUE #40: Feature request - dynamic schema name in safe mode
It would be useful to be able to use a dynamic schema name as follows:

```js
const sql = SQL`SELECT * FROM ${schema}.users WHERE user_id = ${userId}`
```

At the moment, this will fail (e.g. when used with node-postgres), as `schema` is turned in to a parameter in a prepared statement, and, from what I understand, postgres does not support variable schema names in prepared statements.

A workaround is to do this:

```js
const sql = SQL`SELECT * FROM `
sql.append(SQL`${schema}.users `, { unsafe: true })
sql.append(SQL`WHERE user_id = ${userId}`)
```

Switching in and out of unsafe mode is not ideal.

If it can be achieved without risk of introducing SQL injection vulnerabilities, it would be good if this library could handle it and convert:

```js
const schema = 'schema_name'
const userId = 1
const sql = SQL`SELECT * FROM ${schema}.users  WHERE user_id = ${userId}`
```

in to:

```js
{
  text: 'SELECT * FROM "schema_name".users WHERE user_id = ${1}',
  values: [1]
}
```
 
(note that `schema_name` can be wrapped with double quotes in the text, which may help to ensure no SQL injection vulnerabilities)

Alternatively, perhaps there is another way to achieve this, in which case it would be good to document it.

### PR DESCRIPTION
**Pull request by simoneb**
Adds two static functions `unsafe` and `quoteIdent` to interpolate special parameters which are injected as-is or quoted with double quotes, respectively.

Closes #40.

See [new tests](https://github.com/nearform/sql/pull/41/files#diff-cb81f9000da25483d56c7148d0e80ad540322b2391757d6026f37815dc9407edR370) for syntax.

### CODE REVIEW

**Comment by p16** [comment]
Added some more tests to add scenarios in which the new `safe` function may be used.

File: SQL.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
 'use strict'
 const inspect = Symbol.for('nodejs.util.inspect.custom')
+const unsafe = Symbol.for('unsafe')
```
Comment (mcollina) [inline SQL.js:3]:
```suggestion
const unsafe = Symbol('unsafe')
```

Let's not pollute the global registry with this.

File: SQL.test.js [hunk 2]
```diff
@@ -194,7 +195,7 @@ test('SQL helper - build complex query with append and glue', t => {
   )
   t.equal(
     sql.debug,
-    'TEST QUERY glue pieces FROM v1 = \'v1\' , v2 = \'v2\' , v3 = \'v3\' , v4 = \'v4\' , v5 = \'v5\' WHERE v6 = \'v6\' AND v7 = \'v7\''
+    "TEST QUERY glue pieces FROM v1 = 'v1' , v2 = 'v2' , v3 = 'v3' , v4 = 'v4' , v5 = 'v5' WHERE v6 = 'v6' AND v7 = 'v7'"
```
Comment (mcollina) [inline SQL.test.js:198]: why was this changed? I would not expect it to be.

**Review by mcollina — COMMENTED** [review]
The code is good for me. Docs are missing.

**Comment by mcollina** [comment]
Please make sure to run the security checks for sql injections.

File: SQL.test.js (hunk 2 above)
Comment (simoneb) [reply SQL.test.js:198]: because standard was updated and it's standard who decides how quotes are used within strings, which changed across standard versions

**Comment by simoneb** [comment]
@mcollina I have updated the sqlmap test so that the queries executed use the new unsafe function. If that is what you mean by security tests I guess we're covered.

File: SQL.js [hunk 3]
```diff
@@ -44,25 +47,40 @@ class SqlStatement {
...
+    const values = [...this._values]
 
     for (let i = 1; i < this.strings.length; i++) {
-      let delimiter = '?'
-      if (type === 'pg') {
-        delimiter = '$' + i
+      if (values[i - 1 + valueOffset] && values[i - 1 + valueOffset][unsafe]) {
+        text += `${QUOTE}${values[i - 1 + valueOffset].value}${QUOTE}${
+          this.strings[i]
+        }`
```
Comment (paolochiodi) [inline SQL.js:57]: Why are we adding quotes to unsafe values?

I would have two different functions:
- `unsafe` that just mark input to be inserted without changes
- `quoteIdent` that add the quotes 

`quoteIdent` may be something like:
```js
module.exports.quoteIdent = value => unsafe(`${QUOTE}${value}${QUOTE}`)
```

Otherwise I'd find the name `unsafe` misleading.

Comment (simoneb) [reply SQL.js:57]: that's a good point. if we do that, that would raise a question about whether `append(..., { unsafe: true })` has a reason to stay, but I guess we can keep both, at the cost of making the API a little more difficult to understand.

Comment (paolochiodi) [reply SQL.js:57]: Yes, that's a little deviation from current api. 
If that's a problem, we can decide not to expose `unsafe` and have it only used internally.

Comment (simoneb) [reply SQL.js]: @paolochiodi I added quoteIdent based on your suggestion and also exposed unsafe as a function which is an alternative way to using the existing unsafe option

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by simoneb** [comment]
I'll add the docs then merge and release, thanks all for the feedback.

**Review by p16 — APPROVED** [review]
👍
